### DISCUSSION TITLE
Customization variables

### DISCUSSION THREAD

**Discussion by kchanqvq** [neomacs-project/neomacs#66] [Ideas]
I'm thinking about what is the right way to introduce customization variables. Currently there is `default-mixin` in `default-value.lisp`, which is a macro that let me define some slots which takes value from the slot name symbol's symbol property if it is unbound. The rationale was that this allows both global value and buffer-local value, inspired by Emacs's buffer-local variables.

However now I think this is a bit unsatisfactory because:
- They can't be easily rebind dynamically
- As of now I found that slots that holds buffer-local state and slots for configuration seem the be strictly disjoint. Therefore we should remove this and just use `defvar`'s?
- Even if we keep the "slot with default value" idea, we should use slot-definitions rather than slot name symbols?

Related: `define-theme` should have a way to customize these variables? I'm thinking about adding more sections, like `:style`, `:variable`...

@mmontone Do you have any ideas on these?

**Comment by mmontone** [comment]
I've implemented a customization facility for Lem: https://github.com/lem-project/lem/pull/1568

I think we could replicate for Neomacs if you like it. Have a look there and let me know what you think.

**Reply by kchanqvq** [reply]
Thanks for showing this! I'd love to have one in Neomacs!

It seems that this is based on plain-old variables as well. I guess I'll make the move to replace default-value slots with just variables.

**Reply by kchanqvq** [reply]
Once I start moving through the code I become a bit hesitated. There seems to be legitimate case where values associated with slot-definition (rather than a plain-old global variable) make more sense. E.g. Someone might want different `minimum-prefix` for `auto-completion-mode` in different buffers, depending on what modes are enabled...

**Reply by mmontone** [reply]
If you decide you would like to go with customizable class slots, then you could implement with MOP. Use a `:customizable t` slot argument and let the customization facility work with that. Perhaps ...
